core/lib/dal/src/transactions_dal.rs

```diff
@@ -1247,31 +1249,57 @@ impl TransactionsDal<'_, '_> {
         .fetch_all(self.storage.conn())
         .await?;
 
-        assert_eq!(
-            miniblock_data.len(),
-            transactions_by_miniblock.len(),
-            "Not enough miniblock data retrieved"
-        );
-        assert_eq!(
-            prev_hashes.len(),
-            transactions_by_miniblock.len(),
-            "Not enough previous hashes retrieved"
-        );
-
-        Ok(transactions_by_miniblock
+        let prev_miniblock_hashes: HashMap<_, _> = prev_miniblock_hashes
             .into_iter()
-            .zip(miniblock_data)
-            .zip(prev_hashes)
-            .map(
-                |(((number, txs), miniblock_data_row), prev_hash_row)| MiniblockExecutionData {
-                    number,
-                    timestamp: miniblock_data_row.timestamp as u64,
-                    prev_block_hash: H256::from_slice(&prev_hash_row.hash),
-                    virtual_blocks: miniblock_data_row.virtual_blocks as u32,
-                    txs,
-                },
-            )
-            .collect())
+            .map(|row| {
+                (
+                    MiniblockNumber(row.number as u32),
+                    H256::from_slice(&row.hash),
+                )
+            })
+            .collect();
+
+        let mut data = Vec::with_capacity(transactions_by_miniblock.len());
+        let it = transactions_by_miniblock.into_iter().zip(miniblock_data);
+        for ((number, txs), miniblock_row) in it {
+            let prev_miniblock_number = number - 1;
+            let prev_block_hash = match prev_miniblock_hashes.get(&prev_miniblock_number) {
+                Some(hash) => *hash,
+                None => {
+                    // Can occur after snapshot recovery; the first previous miniblock may not be present
```

Comment:
good catch, but I'm curious what we can do (or already doing) so that we don't forget such cases anywhere. Imagine we are adding a new component and ely on the fact that the previous miniblock hash is always present. Is there a way to model data differently so that we don't have to think about it? Probably not. An example would be to store prev miniblock hash in the next miniblock row, but that results in denormalization and is not practical